zkvm/src/vm.rs

```diff
@@ -361,11 +355,31 @@ where
     /// _items... predicate_ **output:_k_** → ø
     fn output(&mut self, k: usize) -> Result<(), VMError> {
         let contract = self.pop_contract(k)?;
-        let output = self.encode_output(contract);
-        self.txlog.push(Entry::Output(output));
+        let mut buf = Vec::with_capacity(contract.min_serialized_length());
+        self.freeze_contract(contract).encode(&mut buf);
+        self.txlog.push(Entry::Output(buf));
         Ok(())
     }
 
+    fn freeze_contract(&mut self, contract: Contract) -> FrozenContract {
+        let frozen_items = contract
+            .payload
+            .into_iter()
+            .map(|i| match i {
+                PortableItem::Data(d) => FrozenItem::Data(d),
+                PortableItem::Value(v) => {
+                    let flv = self.variable_commitments[v.flv.index].closed_commitment();
```

Comment:
do we really need to close the commitment? can we just take `.commitment` as is?